solr/core/src/java/org/apache/solr/parser/SolrQueryParserBase.java

```diff
@@ -600,29 +617,133 @@ protected Query newRegexpQuery(Term regexp) {
-        }
-        DisjunctionMaxQuery dm = new DisjunctionMaxQuery(currPosnClauses, 0.0f);
-        return dm;
-      case AS_DISTINCT_TERMS:
-        BooleanQuery.Builder builder = new BooleanQuery.Builder();
-        for (Term term : terms) {
-          builder.add(newTermQuery(term), BooleanClause.Occur.SHOULD);
+      case PICK_BEST: {
+        List<Query> synonymQueries = getSynonymQueries(field, attributes);
+        return new DisjunctionMaxQuery(synonymQueries, 0.0f);
+      }
+      case AS_DISTINCT_TERMS: {
+        List<Query> synonymQueries = getSynonymQueries(field, attributes);
+        return buildBooleanQuery(synonymQueries);
+      }
+      case AS_SAME_TERM:{
+        SynonymQuery.Builder builder = new SynonymQuery.Builder(field);
+        for (int i = 0; i < attributes.length; i++) {
+          TermToBytesRefAttribute termAttribute = attributes[i].getAttribute(TermToBytesRefAttribute.class);
+          float payloadBoost = getDecodedPayload(attributes[i]);
+          if (isAcceptableBoost(payloadBoost)) {
+            builder.addTerm(new Term(field, termAttribute.getBytesRef()), payloadBoost);
+          } else {
+            builder.addTerm(new Term(field, termAttribute.getBytesRef()));
+          }
         }
         return builder.build();
-      case AS_SAME_TERM:
-        return super.newSynonymQuery(terms);
+      }
       default:
         throw new AssertionError("unrecognized synonymQueryStyle passed when creating newSynonymQuery");
     }
   }
 
+  private List<Query> getSynonymQueries(String field, AttributeSource[] attributes) {
+    List<Query> synonymQueries = new ArrayList<>(attributes.length);
+    for (int i = 0; i < attributes.length; i++) {
+      TermToBytesRefAttribute termAttribute = attributes[i].getAttribute(TermToBytesRefAttribute.class);
+      Query synonymQuery = new TermQuery(new Term(field, termAttribute.getBytesRef()));
+      synonymQueries.add(getBoostedQueryByPayload(new AttributeSource[]{attributes[i]}, synonymQuery));
+    }
+    return synonymQueries;
+  }
+  
+  private Query getBoostedQueryByPayload(AttributeSource[] attributes, Query query) {
+    float payloadBoost = 0f;
+    for (int i = 0; i < attributes.length; i++) {
+      payloadBoost = getDecodedPayload(attributes[i]);
+    }
+    if (isAcceptableBoost(payloadBoost)) {
+      return new BoostQuery(query, payloadBoost);
+    }
+    return query;
+  }
+  
+  private SpanQuery getBoostedQueryByPayload(AttributeSource[] attributes, SpanQuery query) {
+    float payloadBoost = 0f;
+    for (int i = 0; i < attributes.length; i++) {
+      payloadBoost = getDecodedPayload(attributes[i]);
+    }
+    if (isAcceptableBoost(payloadBoost)) {
+      return new SpanBoostQuery(query, payloadBoost);
+    }
+    return query;
+  }
+
+  private float getDecodedPayload(AttributeSource attribute) {
+    float payloadBoost = 0f;
+    PayloadAttributeImpl payloadAttribute = attribute.getAttributeImpl(PayloadAttributeImpl.class);
```

Comment:
You can just use `PayloadAttribute` here?  The implementation is, well, an implementation detail :)

Reply:
definitely make sense!